lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,279 @@
+      continue;
+    }
+    if (!N.hasOneUse()) {
+      RETURN_ERR("Expected constant operation");
+    }
+    auto *SN = dyn_cast<SaveNode>(N.getUsers().begin()->getUser());
+    if (SN && SN->getPlaceholder() == &N) {
+      continue;
+    }
+    RETURN_ERR("Expected constant operation");
+  }
+  return llvm::Error::success();
+}
+
+} // namespace
+
+llvm::Error glow::executeConstantFunction(Backend &backend, Function &F,
+                                          PlaceholderBindings &bindings,
+                                          CompilationContext &cctx) {
+// Perform the checks in DEBUG builds only.
+#ifndef NDEBUG
+  RETURN_IF_ERR(verifyConstantFunction(backend, F));
+#endif
+  auto compiledF = compile(backend, F, cctx);
+  run(backend, *compiledF, bindings);
+  return llvm::Error::success();
+}
+
+/// Perform constant folding in the function \p F . Any non-trivial node (i.e.
+/// not a constant or a splat) that can be computed at compile-time is going to
+/// be computed at compile-time. \returns true if any foldings were performed.
+bool glow::ConstantFold::run(Function *F) {
+  LOG_SCOPE(F->getLogContext(), "glow::constantFold")
+  CompilationContext cctx;
+  // Graph optimizations may be required to e.g. perform lowering of some nodes
+  // that are not supported natively.
+  cctx.optimizationOpts.enableGraphOptz = true;
+  // Do not recursively call constant folding.
+  cctx.optimizationOpts.enableConstantFolding = false;
+  cctx.backendOpts.collectConstants = true;
+  bool changed = false;
+  // Backend to be used for compile-time computations.
+  std::unique_ptr<Backend> backend(createBackend("Interpreter"));
+  // Set of constant operations to be computed.
+  llvm::SmallVector<Node *, 64> worklist;
+  // Traverse nodes in post-order, so that children are seen before parents.
+  GraphPostOrderVisitor postOrderVisitor(*F);
+  auto nodes = postOrderVisitor.getPostOrder();
+  // Collect all non-trivial constant operations.
+  for (auto *N : nodes) {
+    // Skip trivial nodes/operations that do not require any constant
+    // computations.
+    if (llvm::isa<Storage>(N) || llvm::isa<Constant>(N) ||
+        llvm::isa<SplatNode>(N)) {
+      continue;
+    }
+    // Skip nodes that are not constant operations.
+    if (!isConstantOperation(N, *backend)) {
+      continue;
+    }
+    // Add only a constant operation node whose value is used by at least
+    // one non constant-operation node, because no other bigger constant
+    // operation containing the current node can completely replace the result
+    // of its computation. Doing this check allows for performing a smaller
+    // number of evaluateConstantOperation calls later and thus reduces the
+    // overhead.
+    if (!hasNonConstantOperationUser(N, *backend)) {
+      continue;
+    }
+    // Compute the constant value of the node.
+    std::vector<Constant *> constResults =
+        evaluateConstantOperation(*backend, cctx, N);
```

Comment:
NBD, just a thought: with the worklist removal, I'm wondering if we could just call into `glow::constantFold(Node *N)` instead of replicating some of this logic. I see we've already checked above `isConstantOperation()` so as is there'd be some redundant checks.